### DISCUSSION TITLE
Solokeys

### DISCUSSION THREAD

**Discussion by gibait** [dorssel/usbipd-win#309] [Q&A]
Hello, first of all thank you for this project!
I can't seem to make my solo key work with Windows 10. I have kernel `5.10.60.1-microsoft-standard-WSL2` installed, usbipd `usbipd-win 2.1.0` version and i followed every step in the wiki. usbipd shows:
```
C:\WINDOWS\system32>usbipd wsl list
BUSID  DEVICE                                                        STATE
1-1    Logitech USB Input Device, Dispositivo di input USB           Not attached
1-2    Dispositivo di input USB                                      Attached - Ubuntu
1-3    Alcor Micro USB Smart Card Reader                             Not attached
1-7    Intel(R) Wireless Bluetooth(R)                                Not attached
1-8    Integrated Camera                                             Not attached
1-9    Synaptics WBDI - SGX                                          Not attached
1-15   Realtek USB 3.0 Card Reader                                   Not attached
```
And lsusb on WSL:
```
giba@Thinkpad-T480:~$ lsusb  
Bus 002 Device 001: ID 1d6b:0003 Linux Foundation 3.0 root hub
Bus 001 Device 007: ID 0483:a2ca STMicroelectronics Solo 4.1.5
Bus 001 Device 001: ID 1d6b:0002 Linux Foundation 2.0 root hub
```
So, device is connected if i understood properly. dmesg output is ok also:
```
[ 4014.426650] usb 1-1: new full-speed USB device number 7 using vhci_hcd
[ 4014.506842] vhci_hcd: vhci_device speed not set
[ 4014.576810] usb 1-1: SetAddress Request (7) to port 0
[ 4014.612352] usb 1-1: New USB device found, idVendor=0483, idProduct=a2ca, bcdDevice= 1.00
[ 4014.612354] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[ 4014.612355] usb 1-1: Product: Solo 4.1.5
[ 4014.612356] usb 1-1: Manufacturer: SoloKeys
[ 4014.612356] usb 1-1: SerialNumber: 207331B9414E
```
udev rules:
```
giba@Thinkpad-T480:~$ cat /etc/udev/rules.d/70-solokeys-access.rules
# Notify ModemManager this device should be ignored
ACTION!="add|change|move", GOTO="mm_usb_device_blacklist_end"
SUBSYSTEM!="usb", GOTO="mm_usb_device_blacklist_end"
ENV{DEVTYPE}!="usb_device",  GOTO="mm_usb_device_blacklist_end"

ATTRS{idVendor}=="0483", ATTRS{idProduct}=="a2ca", ENV{ID_MM_DEVICE_IGNORE}="1"

LABEL="mm_usb_device_blacklist_end"


# Solo bootloader + firmware access
SUBSYSTEM=="hidraw", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="a2ca", TAG+="uaccess"
SUBSYSTEM=="tty", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="a2ca", TAG+="uaccess"

# ST DFU access
SUBSYSTEM=="usb", ATTRS{idVendor}=="0483", ATTRS{idProduct}=="df11", TAG+="uaccess"

# U2F Zero
SUBSYSTEM=="hidraw", ATTRS{idVendor}=="10c4", ATTRS{idProduct}=="8acf", TAG+="uaccess"
```
drivers:
```
giba@Thinkpad-T480:~$ ls -al /sys/bus/usb/drivers/usbhid
total 0
drwxr-xr-x 2 root root    0 Mar 16 08:40 .
drwxr-xr-x 8 root root    0 Mar 16 08:40 ..
lrwxrwxrwx 1 root root    0 Mar 16 09:55 1-1:1.0 -> ../../../../devices/platform/vhci_hcd.0/usb1/1-1/1-1:1.0
--w------- 1 root root 4096 Mar 16 08:40 bind
lrwxrwxrwx 1 root root    0 Mar 16 08:40 module -> ../../../../module/usbhid
-rw-r--r-- 1 root root 4096 Mar 16 08:40 new_id
-rw-r--r-- 1 root root 4096 Mar 16 08:40 remove_id
--w------- 1 root root 4096 Mar 16 08:40 uevent
--w------- 1 root root 4096 Mar 16 08:40 unbind
```
Kernel recompilation should not be necessary since kernel version is 5.10.60.1 right? What am i missing here?

**Comment by dorssel** [comment]
@gibait 
Everything you posted looks correct. What exactly is the problem you are having?

**Reply by gibait** [reply]
I don't see any output on `/dev/console` for example neither can i interact using `gpg --card-status`.
`solo1-cli` client, also, does not recognize it. It seems that i can't interact with it.
Also, should i not be able to see any `/dev/hidraw` or `/dev/solo`?

~~Maybe i found the issue: the module `/sys/class/hidraw`, necessary for working as stated in [#337](https://github.com/solokeys/solo1/issues/377#issuecomment-589298494l) is not present. Shouldn't it have been implemented with the latest version of the WSL kernel? This is out of scope for usbipd maybe. This means that i will need to recompile kernel?~~

After a reboot, dmesg now gives this:
```[  511.228808] vhci_hcd vhci_hcd.0: pdev(0) rhport(0) sockfd(3)
[  511.228810] vhci_hcd vhci_hcd.0: devid(65538) speed(2) speed_str(full-speed)
[  511.228852] vhci_hcd vhci_hcd.0: Device attached
[  511.504973] vhci_hcd: vhci_device speed not set
[  511.575243] usb 1-1: new full-speed USB device number 5 using vhci_hcd
[  511.655093] vhci_hcd: vhci_device speed not set
[  511.725036] usb 1-1: SetAddress Request (5) to port 0
[  511.760533] usb 1-1: New USB device found, idVendor=0483, idProduct=a2ca, bcdDevice= 1.00
[  511.760534] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[  511.760536] usb 1-1: Product: Solo 4.1.5
[  511.760536] usb 1-1: Manufacturer: SoloKeys
[  511.760537] usb 1-1: SerialNumber: 207331B9414E
[  511.763415] hid-generic 0003:0483:A2CA.0004: device has no listeners, quitting
```

**Reply by dorssel** [reply]
It looks indeed like some drivers are missing. I guess the solokey is similar to a yubikey? Note that WSL only has a very limited set of drivers. Considering that WSL is essentially "headless", I can imagine the HID drivers are not present. You will probably have to compile your own kernel and enable the right CONFIG_XXX.

**Reply by gibait** [reply]
Yes, solokey is similar to yubikey. The strange thing is that after rebooting it had the problem, before it seemed to work. I will try and keep this updated, maybe it will be useful for someone else as well. Thanks

**Reply by aseigler** [reply]
The [default WSL2 kernel config](https://github.com/microsoft/WSL2-Linux-Kernel/blob/linux-msft-wsl-5.10.y/Microsoft/config-wsl) does not have `CONFIG_HIDRAW` set, which I believe is at least part of the problem.  Recompiled with that on, I get 

```
[   12.324391] usb 1-1: new full-speed USB device number 2 using vhci_hcd
[   12.404431] vhci_hcd: vhci_device speed not set
[   12.474556] usb 1-1: SetAddress Request (2) to port 0
[   12.512399] usb 1-1: New USB device found, idVendor=0483, idProduct=a2ca, bcdDevice= 1.00
[   12.512403] usb 1-1: New USB device strings: Mfr=1, Product=2, SerialNumber=3
[   12.512406] usb 1-1: Product: Solo 4.1.0
[   12.512409] usb 1-1: Manufacturer: SoloKeys
[   12.512411] usb 1-1: SerialNumber: 209933B9344B
[   12.518078] hid-generic 0003:0483:A2CA.0001: hiddev96,hidraw0: USB HID v1.11 Device [SoloKeys Solo 4.1.0] on usb-vhci_hcd.0-1/input0
```
But `solo ls` just reports `:: Solos`, so something else seems missing.

**Reply by gibait** [reply]
I tried recompiling it too with the following config:
```
# HID support
#
CONFIG_HID=y
# CONFIG_HID_BATTERY_STRENGTH is not set
CONFIG_HIDRAW=y
CONFIG_UHID=y
CONFIG_HID_GENERIC=y
```
Yet, dmesg still warns me of `device has no listeners, quitting`. 
Did you use solokey udev rules they provide? 
Also can you show me your your `/sys/bus/usb/drivers/usbhid` folder content?

EDIT:
Rebooting has worked for me, now i even see with `solo ls`, only for root user though, must be udev rules fault.
```
(venv) root@Thinkpad-T480:/home/giba/solo1-cli# solo ls
THIS COMMAND SHOULD NOT BE RUN AS ROOT!

Please install udev rules and run `solo` as regular user (without sudo).
For more information, see: https://docs.solokeys.io/solo/udev
:: Solos
207331B9414E: SoloKeys Solo 4.1.5
```

**Reply by aseigler** [reply]
I basically did the same thing and have the same results

```
Please install udev rules and run `solo` as regular user (without sudo).
For more information, see: https://docs.solokeys.io/solo/udev
:: Solos
209933B9344B: SoloKeys Solo 4.1.0
```

Agreed, gotta be some sort of udev failure at this point.  I'm using [this set](https://raw.githubusercontent.com/Yubico/libu2f-host/master/70-u2f.rules) along with [this set](https://raw.githubusercontent.com/solokeys/solo1/master/udev/71-solokeys-symlinks.rules) and none of it appears to be working.  Progress, though.

**Reply by aseigler** [reply]
Installing this got me the rest of the way through: https://github.com/arkane-systems/genie

**Reply by gibait** [reply]
I only used solo [setup](https://github.com/solokeys/solo1/tree/master/udev) and it worked. Also restarted `udev` service
